cirq-core/cirq/work/collector.py

```diff
@@ -163,53 +165,56 @@ async def collect_async(
         Returns:
             The collector's result after all desired samples have been
             collected.
-
-        See Also:
-            Python 3 documentation "Coroutines and Tasks"
-            https://docs.python.org/3/library/asyncio-task.html
         """
-        pool = work_pool.CompletionOrderedAsyncWorkPool()
+        results = duet.AsyncCollector[Tuple[CircuitSampleJob, 'cirq.Result']]()
+        job_error = None
+        running_jobs = 0
         queued_jobs: List[CircuitSampleJob] = []
         remaining_samples = np.infty if max_total_samples is None else max_total_samples
 
-        async def _start_async_job(job):
-            return job, await sampler.run_async(job.circuit, repetitions=job.repetitions)
+        async def run_job(job):
+            nonlocal job_error
+            try:
+                result = await sampler.run_async(job.circuit, repetitions=job.repetitions)
+            except Exception as error:
+                if not job_error:
+                    results.error(error)
+                    job_error = error
+            else:
```

Comment:
what is try .. else? why not put results.add inside the try?

Reply:
The `else` clause executes if the `try` block does not raise any exceptions. I wanted to make clear that the `except` clause is _only_ catching exceptions from the await call. If `results.add` were part of the `try` clause then the except would also handle errors from that; `results.add` shouldn't raise errors, but this is just to be safe.

Reply:
thanks for explaining